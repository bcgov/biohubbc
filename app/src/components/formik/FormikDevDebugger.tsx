import { Button } from '@mui/material';
import { useFormikContext } from 'formik';
import { useState } from 'react';

/**
 * Renders formik form values. Used for manual testing purposes
 * Only renders display button when in development.
 * Can provide a custom payload / object for additional testing.
 *
 * @params {FormikDevDebugger}
 * @returns {*}
 *
 **/

interface FormikDevDebuggerProps {
  custom_payload?: any;
}
const FormikDevDebugger = ({ custom_payload }: FormikDevDebuggerProps) => {
  const { values, errors, touched } = useFormikContext();
  const [showFormDebugger, setShowDebugger] = useState(false);
  return (
    <>
      {/* Only render the button in Development */}
      {process.env.NODE_ENV === 'development' ? (
        <Button onClick={() => setShowDebugger((d) => !d)}>Formik Debugger (Only Dev)</Button>
      ) : null}
      {showFormDebugger ? (
        <pre>
          {JSON.stringify(
            { FORMIK_ERRORS: errors, FORMIK_VALUES: values, FORMIK_TOUCHED: touched, CUSTOM_PAYLOAD: custom_payload },
            null,
            2
          )}
        </pre>
      ) : null}
    </>
  );
};

export default FormikDevDebugger;
