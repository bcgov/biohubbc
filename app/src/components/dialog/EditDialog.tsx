import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import useTheme from '@material-ui/core/styles/useTheme';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import { Formik, FormikValues } from 'formik';
import React, { PropsWithChildren } from 'react';

export interface IEditDialogComponentProps<T> {
  element: any;
  initialValues: T;
  validationSchema: any;
}

export interface IEditDialogProps<T> {
  /**
   * The dialog window title text.
   *
   * @type {string}
   * @memberof IEditDialogProps
   */
  dialogTitle: string;

  /**
   * The label of the `onSave` button.
   *
   * Defaults to `Save Changes` if not specified.
   *
   * @type {string}
   * @memberof IEditDialogProps
   */
  dialogSaveButtonLabel?: string;
  /**
   * Set to `true` to open the dialog, `false` to close the dialog.
   *
   * @type {boolean}
   * @memberof IEditDialogProps
   */
  open: boolean;

  /**
   * @type {IEditDialogComponentProps}
   * @memberof IEditDialogProps
   */
  component: IEditDialogComponentProps<T>;

  /**
   * Error message to display when an error exists
   */
  dialogError?: string;

  /**
   * Callback fired if the 'No' button is clicked.
   *
   * @memberof IEditDialogProps
   */
  onCancel: () => void;
  /**
   * Callback fired if the 'Yes' button is clicked.
   *
   * @memberof IEditDialogProps
   */
  onSave: (values: T) => void;
}

/**
 * A dialog for displaying a component for editing purposes and giving the user the option to say
 * `Yes`(Save) or `No`.
 *
 * @template T
 * @param {PropsWithChildren<IEditDialogProps<T>>} props
 * @return {*}
 */
export const EditDialog = <T extends FormikValues>(props: PropsWithChildren<IEditDialogProps<T>>) => {
  const theme = useTheme();

  const fullScreen = useMediaQuery(theme.breakpoints.down('sm'));

  if (!props.open) {
    return <></>;
  }

  return (
    <Formik
      initialValues={props.component.initialValues}
      enableReinitialize={true}
      validationSchema={props.component.validationSchema}
      validateOnBlur={true}
      validateOnChange={false}
      onSubmit={(values) => {
        props.onSave(values);
      }}>
      {(formikProps) => (
        <Dialog
          fullScreen={fullScreen}
          maxWidth="xl"
          open={props.open}
          aria-labelledby="edit-dialog-title"
          aria-describedby="edit-dialog-description">
          <DialogTitle id="edit-dialog-title">{props.dialogTitle}</DialogTitle>
          <DialogContent>{props.component.element}</DialogContent>
          <DialogActions>
            <Button onClick={formikProps.submitForm} color="primary" variant="contained" autoFocus>
              {props.dialogSaveButtonLabel || 'Save Changes'}
            </Button>
            <Button onClick={props.onCancel} color="primary" variant="outlined">
              Cancel
            </Button>
          </DialogActions>
          {props.dialogError && <DialogContent>{props.dialogError}</DialogContent>}
        </Dialog>
      )}
    </Formik>
  );
};

export default EditDialog;
