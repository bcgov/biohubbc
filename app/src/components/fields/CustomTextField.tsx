import TextField, { TextFieldProps } from '@material-ui/core/TextField';
import { useFormikContext } from 'formik';
import React from 'react';

export interface ICustomTextField {
  label: string;
  name: string;
  other?: TextFieldProps;
}

const CustomTextField: React.FC<ICustomTextField> = (props) => {
  const { touched, errors, values, handleChange } = useFormikContext<object>();
  const { name, label, other } = props;

  return (
    <TextField
      name={name}
      label={label}
      id={name}
      onChange={handleChange}
      variant="outlined"
      value={values[name]}
      fullWidth={true}
      error={touched[name] && Boolean(errors[name])}
      helperText={touched[name] && errors[name]}
      {...other}
    />
  );
};

export default CustomTextField;
