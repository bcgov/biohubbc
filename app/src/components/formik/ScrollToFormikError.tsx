import Snackbar from '@material-ui/core/Snackbar';
import Alert from '@material-ui/lab/Alert';
import { useFormikContext } from 'formik';
import { IGetProjectForViewResponse } from 'interfaces/useProjectApi.interface';
import React, { useEffect, useState } from 'react';

export interface IScrollToFormikErrorProps {
  /**
   * An ordered list of field names, which informs which field this component will scroll to when multiple fields are
   * in error.
   *
   * Note: A regex is required if the field name has dynamic components (ie: a field with an array of objects where
   * part of the field name is its index in the array field).
   *
   * @type {((string | RegExp)[])}
   * @memberof IScrollToFormikErrorProps
   */
  fieldOrder: (string | RegExp)[];
}

export const ScrollToFormikError: React.FC<IScrollToFormikErrorProps> = (props) => {
  const formikProps = useFormikContext<IGetProjectForViewResponse>();
  const { errors, submitCount } = formikProps;
  const [openSnackbar, setOpenSnackbar] = useState({ open: false, msg: '' });

  useEffect(() => {
    const showSnackBar = (message: string) => {
      setOpenSnackbar({ open: true, msg: message });
    };

    const getAllFieldErrorNames = (obj: object, prefix = '', result: string[] = []) => {
      Object.keys(obj).forEach((key) => {
        const value = obj[key];
        if (!value) return;

        key = Number(key) || key === '0' ? `[${key}]` : key;

        const nextKey = prefix ? `${prefix}.${key}` : key;

        if (typeof value === 'object') {
          getAllFieldErrorNames(value, nextKey, result);
        } else {
          result.push(nextKey);
        }
      });
      return result;
    };

    const getFirstErrorField = (errorArray: string[]): string | undefined => {
      for (const listError of props.fieldOrder) {
        for (const trueError of errorArray) {
          if (trueError.match(listError) || listError === trueError) {
            return trueError;
          }
        }
      }
    };

    const fieldErrorNames = getAllFieldErrorNames(errors);

    const topFieldError = getFirstErrorField(fieldErrorNames);

    if (!topFieldError) {
      return;
    }

    showSnackBar(`Missing one or more required fields.`);

    const errorElement = document.getElementsByName(topFieldError);

    if (errorElement.length <= 0) {
      return;
    }

    errorElement[0].scrollIntoView({ behavior: 'smooth', block: 'center' });

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [errors, submitCount]);

  const closeSnackBar = () =>
    setOpenSnackbar(() => {
      return { open: false, msg: '' };
    });

  return (
    <>
      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center'
        }}
        open={openSnackbar.open}
        onClose={closeSnackBar}>
        <Alert elevation={4} severity="error" variant="filled" onClose={closeSnackBar}>
          {openSnackbar.msg}
        </Alert>
      </Snackbar>
    </>
  );
};
