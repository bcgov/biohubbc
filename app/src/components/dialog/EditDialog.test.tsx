import { fireEvent, render, waitFor } from '@testing-library/react';
import EditDialog from 'components/dialog/EditDialog';
import CustomTextField from 'components/fields/CustomTextField';
import { useFormikContext } from 'formik';
import React from 'react';
import yup from 'utils/YupSchema';

export interface ISampleFormikFormProps {
  testField: string;
}

export const SampleFormikFormYupSchema = yup.object().shape({
  testField: yup
    .string()
    .max(3000, 'Cannot exceed 3000 characters')
    .required('You must provide a test field for the project')
});

const SampleFormikForm = () => {
  const formikProps = useFormikContext<ISampleFormikFormProps>();

  const { handleSubmit } = formikProps;

  return (
    <form onSubmit={handleSubmit}>
      <CustomTextField name="testField" label="Test Field" other={{ multiline: true, required: true, rows: 4 }} />
    </form>
  );
};

const handleOnSave = jest.fn();
const handleOnCancel = jest.fn();

const renderContainer = ({
  testFieldValue,
  dialogError,
  open = true
}: {
  testFieldValue: string;
  dialogError?: string;
  open?: boolean;
}) => {
  return render(
    <div id="root">
      <EditDialog
        dialogTitle="This is dialog title"
        dialogError={dialogError || undefined}
        open={open}
        component={{
          element: <SampleFormikForm />,
          initialValues: { testField: testFieldValue },
          validationSchema: SampleFormikFormYupSchema
        }}
        onCancel={handleOnCancel}
        onSave={handleOnSave}
      />
    </div>
  );
};

describe('EditDialog', () => {
  it('matches the snapshot when not open', () => {
    const { baseElement } = renderContainer({ testFieldValue: 'this is a test', open: false });

    expect(baseElement).toMatchSnapshot();
  });

  it('matches the snapshot when open, with no error message', () => {
    const { baseElement } = renderContainer({ testFieldValue: 'this is a test' });

    expect(baseElement).toMatchSnapshot();
  });

  it('matches snapshot when open, with error message', () => {
    const { baseElement } = renderContainer({ testFieldValue: 'this is a test', dialogError: 'This is an error' });

    expect(baseElement).toMatchSnapshot();
  });

  it('calls the onSave prop when `Save Changes` button is clicked', async () => {
    const { findByText, getByLabelText } = renderContainer({ testFieldValue: 'initial value' });

    const textField = await getByLabelText('Test Field', { exact: false });

    fireEvent.change(textField, { target: { value: 'updated value' } });

    const saveChangesButton = await findByText('Save Changes', { exact: false });

    fireEvent.click(saveChangesButton);

    await waitFor(() => {
      expect(handleOnSave).toHaveBeenCalledTimes(1);
      expect(handleOnSave).toHaveBeenCalledWith({
        testField: 'updated value'
      });
    });
  });

  it('calls the onCancel prop when `Cancel` button is clicked', async () => {
    const { findByText } = renderContainer({ testFieldValue: 'this is a test' });

    const cancelButton = await findByText('Cancel', { exact: false });

    fireEvent.click(cancelButton);

    await waitFor(() => {
      expect(handleOnCancel).toHaveBeenCalledTimes(1);
    });
  });
});
