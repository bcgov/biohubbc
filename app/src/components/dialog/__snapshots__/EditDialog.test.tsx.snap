// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditDialog matches snapshot when open, with error message 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      id="root"
    />
  </div>
  <div
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-describedby="edit-dialog-description"
        aria-labelledby="edit-dialog-title"
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthXl MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
          id="edit-dialog-title"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            This is dialog title
          </h2>
        </div>
        <div
          class="MuiDialogContent-root"
        >
          <form>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
                data-shrink="true"
                for="testField"
                id="testField-label"
              >
                Test Field
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                  id="testField"
                  name="testField"
                  required=""
                  rows="4"
                >
                  this is a test
                </textarea>
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-7 PrivateNotchedOutline-legendNotched-8"
                  >
                    <span>
                      Test Field
                       *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </form>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-focusVisible Mui-focusVisible"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Save Changes
            </span>
            <span
              class="MuiTouchRipple-root"
            >
              <span
                class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
                style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
              >
                <span
                  class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
                />
              </span>
            </span>
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Cancel
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiDialogContent-root"
        >
          This is an error
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`EditDialog matches the snapshot when not open 1`] = `
<body>
  <div>
    <div
      id="root"
    />
  </div>
</body>
`;

exports[`EditDialog matches the snapshot when open, with no error message 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      id="root"
    />
  </div>
  <div
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-describedby="edit-dialog-description"
        aria-labelledby="edit-dialog-title"
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthXl MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
          id="edit-dialog-title"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            This is dialog title
          </h2>
        </div>
        <div
          class="MuiDialogContent-root"
        >
          <form>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
                data-shrink="true"
                for="testField"
                id="testField-label"
              >
                Test Field
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                  id="testField"
                  name="testField"
                  required=""
                  rows="4"
                >
                  this is a test
                </textarea>
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
                  >
                    <span>
                      Test Field
                       *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </form>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-focusVisible Mui-focusVisible"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Save Changes
            </span>
            <span
              class="MuiTouchRipple-root"
            >
              <span
                class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
                style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
              >
                <span
                  class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
                />
              </span>
            </span>
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Cancel
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
